wire-ios-sync-engine/Source/Calling/MLSConferenceStaleParticipantsRemover.swift

```diff
@@ -0,0 +1,224 @@
+            }
+        }
+
+        return newAndChanged
+    }
+
+    // MARK: - Helpers
+
+    private func subconversationMembers(for groupID: MLSGroupID) -> [MLSClientID]? {
+        do {
+            return try mlsService.subconversationMembers(for: groupID)
+        } catch {
+            logger.warn("failed to fetch subconversation members: \(String(describing: error))")
+            return nil
+        }
+    }
+
+    private func enqueueRemove(
+        client clientID: MLSClientID,
+        from groupID: MLSGroupID,
+        after duration: TimeInterval
+    ) {
+        do {
+            try timerManager.startTimer(
+                for: clientID,
+                duration: duration,
+                completion: { [weak self] in
+                    guard let self = self else { return }
+
+                    self.remove(
+                        client: clientID,
+                        from: groupID
+                    )
+                }
+            )
+
+            logger.info("started timer for removal of stale participant (clientdID: \(clientID), groupID: \(groupID))")
+        } catch TimerError.timerAlreadyExists {
+            // timer already exists, do nothing
+        } catch {
+            logger.warn("failed to start timer for removal of stale participant (clientdID: \(clientID), groupID: \(groupID)). error: (\(error))")
+        }
+    }
+
+    private func remove(
+        client clientID: MLSClientID,
+        from groupID: MLSGroupID
+    ) {
+        context.perform { [weak self] in
+            guard let `self` = self else { return }
+
+            Task {
+                do {
+                    let subconversationMembers = try self.mlsService.subconversationMembers(for: groupID)
+
+                    guard subconversationMembers.contains(clientID) else {
+                        return
+                    }
+
+                    try await self.mlsService.removeMembersFromConversation(with: [clientID], for: groupID)
+                    self.logger.info("removed stale participant from subconversation (clientID: \(clientID), groupID: \(groupID))")
+                } catch {
+                    self.logger.error("failed to remove stale participant from subconversation: \(String(describing: error))")
+                }
+            }
+        }
+    }
+
+    private func cancelRemoval(for clientID: MLSClientID) {
+        do {
+            try timerManager.cancelTimer(for: clientID)
+            logger.info("canceled removal of participant (\(clientID))")
+        } catch TimerError.timerNotFound {
+            // no timer to cancel, do nothing
+        } catch {
+            logger.warn("failed to cancel removal of participant (\(clientID))")
+        }
+    }
+}
+
+private extension Array where Element == CallParticipant {
+
+    func excludingSelf(in context: NSManagedObjectContext) -> Self {
+        var selfUserID: AVSIdentifier!
+
+        context.performAndWait {
```

Comment:
on which thread is this executed?

Reply:
`excludingSelf` is called from the main thread and `context` is the sync context, so it will be executed there

Reply:
so blocking the mainThread with performAndWait

Reply:
good catch, I missed it. I'll update soon

Reply:
done [6a14183](https://github.com/wireapp/wire-ios-mono/pull/395/commits/6a1418373b73eb0b753ec7b683b640adde9ee17e)